staging/src/k8s.io/client-go/rest/alternative_services.go

```diff
@@ -0,0 +1,367 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rest
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+// AltSvcHandler is an interface for handling AltSvc headers
+type AltSvcHandler interface {
+	// HandleAltSvcHeader is called when an Alt-Svc header is received together with the host of the server
+	HandleAltSvcHeader(altSvcString, host string)
+	BaseURL() url.URL
+}
+
+// NoAltSvcs is an implementation of AltSvcHandler that suppresses warnings.
+type NoAltSvcs struct{}
+
+var _ AltSvcHandler = &NoAltSvcs{}
+
+func (n *NoAltSvcs) HandleAltSvcHeader(altSvcString, host string) {}
+func (n *NoAltSvcs) BaseURL() url.URL {
```

Comment:
This is called everytime we create a NewRequest, it returns an url from the cache (it does some validations)